### ISSUE TITLE
Code generation changes cause CPU aborts with use of format_args!()

### ISSUE THREAD

**Issue by berkus** [rust-lang/rust#102849]
I am using nightly compiler.

It worked well until a point, and now started generating code that GPFs.

The project and branch for testing this is here: https://github.com/metta-systems/vesper/tree/investigate/invalid-code-gen

In the `rust-toolchain` file I have found the exact version which breaks: nightly-2022-08-12 works, while nightly-2022-08-13 started generating something incorrectly. 
I have looked though the generated machine code in the ELF binary and it seems to be exactly the same.

You can test yourself by running `just qemu` with one or another version of nightly.

# NB: UPDATED the code below seems to be misleading, there's other stuff breaking before that, see my comments below.

With 08-13 this happens:

```
----------------
IN:
0x00092a38:  f9400268  ldr      x8, [x19]
0x00092a3c:  a94227e0  ldp      x0, x9, [sp, #0x20]
0x00092a40:  8b141108  add      x8, x8, x20, lsl #4
0x00092a44:  a9400901  ldp      x1, x2, [x8]
0x00092a48:  f9400d28  ldr      x8, [x9, #0x18]
0x00092a4c:  d63f0100  blr      x8                                  ; this is a virtual call in function Console::replace_with()

Taking exception 3 [Prefetch Abort] on CPU 0
...from EL1 to EL1
...with ESR 0x21/0x86000000
...with FAR 0x9599000000000
...with ELR 0x9599000000000
...to EL1 PC 0x96200 PSTATE 0x3c5
----------------
```

The console resides in the BSS section, as it is fully zero-initialised at the start. Since I was working on BSS initialization for my OS I at first suspected something is wrong with my code, so I replaced it with a simple plain-asm version: see [this commit](https://github.com/metta-systems/vesper/commit/5547a7e9f97035d6c0b6c368586a80459a65fd62), however, the behavior persisted and triggered a search for exact rustc version that breaks.

Version compiled with 08-12 works well with both versions of BSS init code (inline rust that calls to memset and an asm version), see [this commit](https://github.com/metta-systems/vesper/commit/eee1d2c8b9b256a50e28a7577169b9f921210865).

Generated code to compare should be in `target/nucleus.bin`

### Version it worked on

nightly-2022-08-12

### Version with regression

Broke on nightly-2022-08-13 and remains so.

**Comment by nagisa** [comment]
Some targeted questions: Is this generally reproducible or with QEMU only?

As a more general note, it may be necessary for you to narrow down the problem further (minimal reproducible code sample would be great), more precisely bisect the offending commit ranges ([cargo-bisect-rustc](https://github.com/rust-lang/cargo-bisect-rustc) can help somewhat), etc.

**Comment by berkus** [comment]
I've only noticed and tested it on QEMU so far, could try on real hw as well, but the fact that it works on the same QEMU with 08-12 means something.

I will try to bisect rustc.

**Comment by nagisa** [comment]
Reproducing this on real hardware would beyond reasonable doubt rule out any possible bugs in QEMU.

**Comment by berkus** [comment]
> Reproducing this on real hardware would beyond reasonable doubt rule out any possible bugs in QEMU.

True. I'll need to restore my hw setup for that though.

**Comment by berkus** [comment]
I've validated this on real Raspberry Pi 4.

Kernel compiled with 08-12 boots fine.
Kernel compiled with 08-13 doesn't.

It will take some time to investigate the details, will need to jump through numerous gdb hoops.

**Comment by berkus** [comment]
Weird note that this triggers only in one specific kernel - the second kernel I build from same sources but that uses slightly different boot process (it relocates itself to higher memory addresses) seems to boot fine even with 08-13.

This smells of an UB.

**Comment by berkus** [comment]
I made something of a reproduction reduced to a simple use of fmt::Write (with a {} in the format string it simply stops working in 08-13 nightly but works in 08-12), will upload it soon.

**Comment by berkus** [comment]
The minimal repro code is here:

- https://github.com/metta-systems/vesper/tree/wip/rustc-102849-repro (note the branch `wip/rustc-102849-repro`)
- `just qemu` should trigger the happy case
- modify rust toolchain to use nightly-2022-08-13 instead of 08-12 to see it failing (not printing anything in semihosting qemu output)
- comment out a marked line in src/main.rs to see it working again

Totally could be something I missed (an esoteric section rustc generates for format_args?), but this is the smallest I've got.

**Comment by thomcc** [comment]
What linker are you using? I've seen this before with mold before, but never tracked it down after learning it didn't repro on other linkers.

**Comment by berkus** [comment]
@thomcc this is using `rust-lld`, as per [target json](https://github.com/metta-systems/vesper/blob/85fe40f603e500518af96d6aa922f7ecfa57f0c5/targets/aarch64-vesper-metta.json#L11-L12).

**Comment by apiraino** [comment]
removing priority label until the context of this issue clarifies

@rustbot label -I-prioritize

**Comment by berkus** [comment]
The issue seems to be fixed in the latest nightly and slightly prior to that, so I will close it.

In case it re-occurs, this is the versions i've tested on my [reference commit](https://github.com/metta-systems/vesper/commit/c06d89a23c5c2b0cd3ed5d270aec10054216ea5a) using `just clean; just qemu`:

# channel=nightly-2022-08-12 "rustc 1.65.0-nightly (20ffea693 2022-08-11)" works
# channel=nightly-2022-08-13 "rustc 1.65.0-nightly (f22819bcc 2022-08-12)" doesn't
# channel=nightly-2023-01-10 "rustc 1.68.0-nightly (3020239de 2023-01-09)" doesn't
# channel=nightly-2023-02-15 "rustc 1.69.0-nightly (0416b1a6f 2023-02-14)" doesn't
# channel=nightly-2023-03-15 "rustc 1.70.0-nightly (171693274 2023-03-14)" doesn't
# channel=nightly-2023-03-25 "rustc 1.70.0-nightly (8be3c2bda 2023-03-24)" doesn't
# channel=nightly-2023-03-26 "rustc 1.70.0-nightly (0c61c7a97 2023-03-25)" works
# channel=nightly-2023-03-27 "rustc 1.70.0-nightly (db0cbc48d 2023-03-26)" works
# channel=nightly-2023-03-30 "rustc 1.70.0-nightly (17c116721 2023-03-29)" works
# channel=nightly-2023-04-10 "rustc 1.70.0-nightly (84dd17b56 2023-04-14)" works
# channel=nightly-2023-07-07 "rustc 1.72.0-nightly (cb80ff132 2023-07-07)" works

So whatever changed, 
* it broke between [20ffea693..f22819bcc](https://github.com/rust-lang/rust/compare/20ffea693..f22819bcc) 
* and got fixed between [8be3c2bda..0c61c7a97](https://github.com/rust-lang/rust/compare/8be3c2bda..0c61c7a97)

Sorry I couldn't be of more detailed help here.

**Comment by mati865** [comment]
There are no definitive proofs but I think it might be related to LLVM.

> it broke between [20ffea693..f22819bcc](https://github.com/rust-lang/rust/compare/20ffea693..f22819bcc)


Upgrade to LLVM 14 happened here.

> and got fixed between [8be3c2bda..0c61c7a97](https://github.com/rust-lang/rust/compare/8be3c2bda..0c61c7a97)

Upgrade to LLVM 15 happened here.

**Comment by berkus** [comment]
This is very probable, but I don't have enough resources to hunt down breaking changes across major LLVM releases. Lets hope it stays fixed)
